// ng2-tag-input/dist/modules/components/helpers/events-actions.js
export function customSeparatorKeys($event) {
    if (this.separatorKeys.indexOf($event.keyCode) >= 0) {
        $event.preventDefault();
        this.addItem();
    }
}
export function backSpaceListener($event) {
    var itemsLength = this.items.length, inputValue = this.inputForm.value.value, isCorrectKey = $event.keyCode === 37 || $event.keyCode === 8;
    if (isCorrectKey && !inputValue && itemsLength) {
        this.tags.last.select.call(this.tags.last);
    }
}
export function addListener(listenerType, action, condition) {
    if (condition === void 0) { condition = true; }
    if (!this.listeners.hasOwnProperty(listenerType)) {
        throw new Error('The event entered may be wrong');
    }
    if (!condition) {
        return;
    }
    this.listeners[listenerType].push(action);
}
function getMatchingItems(value) {
    var _this = this;
    if (!value && !this.showDropdownIfEmpty) {
        return [];
    }
    var itemsMatching = [];
    var items = this.autocompleteItems;
    var lowercaseValue = value.toLowerCase();
    items.forEach(function (item) {
        var condition = item.toLowerCase().indexOf(lowercaseValue) >= 0 && !_this.findItem(item);
        if (condition) {
            itemsMatching.push(item);
        }
    });
    return itemsMatching;
}
export function autoCompleteListener(ev) {
    var value = this.inputForm.value.value;
    var position = this.inputForm.getElementPosition();
    var key = ev.keyCode;
    var itemsMatching = getMatchingItems.call(this, value);
    this.itemsMatching = itemsMatching;
    if (itemsMatching.length || (this.showDropdownIfEmpty && !value)) {
        var focus_1 = key === 40 ? true : false;
        this.dropdown.dropdown.show(position, focus_1);
    }
    if (!itemsMatching.length && this.dropdown.isVisible) {
        this.dropdown.dropdown.hide();
    }
}
export function onAutocompleteItemClicked(item) {
    var _this = this;
    if (!item) {
        return;
    }
    if (this.isTagValid(item.value, true)) {
        this.appendNewTag(item.value);
    }
    this.setInputValue('');
    setTimeout(function () { return _this.inputForm.focus(); }, 0);
    this.dropdown.dropdown.hide();
}
